[Clement Tee] ip

clement0010 opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Add great, echo, exit feature

Commit:
Add todo add, todo list feature

Commit:
Add task class

Commit:
Add mark and unmark todo feature

Commit:
Add todo class

Commit:
Add Deadline class

Commit:
Add Event class

Commit:
Add todo, deadline, event feature

Commit:
Add duke exception class

Commit:
Add empty todo exception class

Commit:
Add duke, empty todo exception handling

Commit:
Add delete todo feature

Commit:
Fix code format

Commit:
Update testing input.txt and expected.txt

Commit:
Add instruction type enum

Commit:
Use instruction type enum

Commit:
Update gitignore

Commit:
Create new save file format

Commit:
Create storage class to handle file saving

Commit:
Use storage class

Commit:
Add datetime formatter to deadline and event class

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Create commands

Commit:
Update test input

Commit:
Create ui file

Commit:
Create messages class

Commit:
Create tasklist class

Commit:
Create parser class

Commit:
Refactor code to follow OOP principle

Commit:
Refactor code base to packages

Commit:
Add junit test

Commit:
Setup for first jar release

Commit:
Add javadoc to parser

Commit:
Use single file import

Commit:
Create find task command

Commit:
Extend find task feature

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge tag 'branch-A-JavaDoc'

Commit:
Merge tag 'branch-A-CodingStandard'

Commit:
Merge tag 'branch-Level-9'

Gernene commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+        if (isValidPath(path)) {
+            return taskList;
+        }
+        try(BufferedReader reader = new BufferedReader(new FileReader(path.toString()))) {
```
I think you're missing a tiny space here.

Gernene commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,107 @@
+public class Parser {
+    private static final String DEFAULT_DATE_FORMAT = "d/MM/yyyy HHmm";
+
+    public static Command parse(String userInput) throws DukeException {
```
This is a total nitpick, but don't forget to add your header comments! xD

kerburenthusiasm commented on src/main/java/duke/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,25 @@
+import duke.task.Todo;
+import duke.ui.TextUi;
+
+public class AddTodoCommand extends Command {
```
Perhaps follow a consistent style in naming classes extending `Command` class?
```suggestion
public class TodoCommand extends Command {
```

Gernene reviewed (commented):
Hi! I couldn't find many coding standard items to fix. I left a few nitpick markers here and there, but otherwise looks good. Nice work!

kerburenthusiasm commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,20 @@
+public abstract class Command {
+    private final boolean isExit;
+
+    public Command (boolean isExit) {
```
Minor change needed here.

```suggestion
    public Command(boolean isExit) {
```

kerburenthusiasm commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,47 @@
+
+    }
+
+    public boolean contains(String keywords) {
```
Perhaps a function name which sounds more like a boolean?
```suggestion
    public boolean hasKeywords(String keywords) {
```

kerburenthusiasm commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,20 @@
+
+public class DukeTest {
+    @Test
+    public void testTaskSizeParser() {
```
Perhaps a more descriptive name for the test function?

```suggestion
    public void testTaskSizeParser_emptyString() {
```

kerburenthusiasm commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,47 @@
+            return "D";
+        } else {
+            return " ";
+        }
```
Perhaps throw an exception for the default case?

```suggestion
        } else {
            throw new DukeException("Unknown Task");
        }
```

kerburenthusiasm commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,59 @@
+        tasks.add(task);
+    }
+
+    public TaskList filterTask(String keywords) {
```
Perhaps a function name which sounds more like a plural as the function is filtering among many tasks?

```suggestion
    public TaskList filterTasks(String keywords) {
```

kerburenthusiasm commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,99 @@
+        }
+    }
+
+    public ArrayList<Task> loadALlTasks() throws DukeException  {
```
Minor mistake found in function name.

```suggestion
    public ArrayList<Task> loadAllTasks() throws DukeException  {
```

kerburenthusiasm commented on src/main/java/duke/command/FindTaskCommand.java:
```diff
@@ -0,0 +1,28 @@
+    public void execute(TaskList tasks, TextUi ui, Storage storage) {
+        try {
+            TaskList filteredTasks = tasks.filterTask(keywords);
+            for (int i = 0; i < filteredTasks.getSize(); i++) {
+                ui.showToUser((i + 1) + ". " + filteredTasks.getTask(i));
+            }
```
Perhaps encapsulate this part of the code into a function to reduce code duplication?

```suggestion
            printFilteredTasks();
```

kerburenthusiasm reviewed (commented):
Overall, the code is readable and strictly follows Java coding standard. However, some of the files requires header comments for the classes and functions. I noted a few minor coding standard violations.

arcornior commented on src/main/java/duke/common/Messages.java:
```diff
@@ -0,0 +1,18 @@
+    public static final String MESSAGE_ERROR_INVALID_COMMAND = "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
+    public static final String MESSAGE_ERROR_EMPTY_TITLE = "☹ OOPS!!! The description of a todo cannot be empty.";
+    public static final String MESSAGE_ERROR_INVALID_DATETIME_FORMAT= "Invalid date time format.";
+}
```
Good job with the constant naming convention!

arcornior commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,107 @@
+
+    public static Command parse(String userInput) throws DukeException {
+        String[] userInputArr = userInput.split(" ", 2);
+        switch (userInputArr[0]) {
+            case "mark":
+                return new MarkTaskCommand(Integer.parseInt(userInputArr[1]) - 1, true);
+            case "unmark":
+                return new MarkTaskCommand(Integer.parseInt(userInputArr[1]) - 1, false);
+            case "delete":
+                return new DeleteCommand(Integer.parseInt(userInputArr[1]) - 1);
+            case "find":
+                userInput = String.join(" ",
+                        Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
+                return new FindTaskCommand(userInput);
+            case "list":
+                return new ListCommand();
+            case "bye":
+                return new ExitCommand();
+            case "todo":
+                return prepareAddTodoCommand(userInputArr[1]);
+            case "event":
+                userInput = String.join(" ",
+                        Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
+                return prepareAddEventCommand(userInput);
+            case "deadline":
+                userInput = String.join(" ",
+                        Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
+                return prepareAddDeadlineCommand(userInput);
+            default:
+                throw new DukeException(Messages.MESSAGE_ERROR_INVALID_COMMAND);
+        }
```
Shouldn't the case indent be on the same level as the switch statement?
I noticed that this issue occurred a few times in your program. Do consider taking note of this.
```suggestion
        switch (userInputArr[0]) {
        case "mark":
            return new MarkTaskCommand(Integer.parseInt(userInputArr[1]) - 1, true);
        case "unmark":
            return new MarkTaskCommand(Integer.parseInt(userInputArr[1]) - 1, false);
        case "delete":
            return new DeleteCommand(Integer.parseInt(userInputArr[1]) - 1);
        case "find":
            userInput = String.join(" ",
                    Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
            return new FindTaskCommand(userInput);
        case "list":
            return new ListCommand();
        case "bye":
            return new ExitCommand();
        case "todo":
            return prepareAddTodoCommand(userInputArr[1]);
        case "event":
            userInput = String.join(" ",
                    Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
            return prepareAddEventCommand(userInput);
        case "deadline":
            userInput = String.join(" ",
                    Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
            return prepareAddDeadlineCommand(userInput);
        default:
            throw new DukeException(Messages.MESSAGE_ERROR_INVALID_COMMAND);
        }
```

arcornior commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @Override
+    public void execute(TaskList tasks, TextUi ui, Storage storage) {
```
Should there be a JavaDocs for this method?

arcornior commented on src/main/java/duke/ui/TextUi.java:
```diff
@@ -0,0 +1,80 @@
+package duke.ui;
+
+import duke.common.Messages;
+import duke.parser.Parser;
+
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.Scanner;
```
Great job separating your import statements!

arcornior reviewed (commented):
LGTM! Just a few nits to fix.

Commit:
Merge branch 'add-gradle-support'

Commit:
Update build gradle configuration file

Commit:
Add checkstyle configuration files

Commit:
Fix style

Commit:
Add javafx configuration file

Commit:
Add resources file image

Commit:
Update all commands and text ui to use javafx

Commit:
Merge branch 'branch-Level-10'

Commit:
Fix code style

Violation of code style: Trailing whitespaces, indentation, variable naming.

The violation of code style decreases readability and maintainability.

Let's,
* update variable naming in tasklist
* remove trailing whitespace 
* fix switch case indentation
* add descriptive variable name

Refer to java coding standard
https://se-education.org/guides/conventions/java/intermediate.html

Commit:
Add assertions to document important assumptions

Commit:
Fix styling and remove redundant imports

Commit:
Use stream in filter task function

Commit:
Configure gradle file to point to launcher

Commit:
Add CI github action workflow

Commit:
Add java docstring for all public methods

clement0010 committed:
Merge branch 'master' into branch-A-CodeQuality

clement0010 committed:
Merge pull request #3 from clement0010/branch-A-Assertions

Add assertions to document important assumptions

clement0010 committed:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Add java docstring and remove trailing white space

clement0010 committed:
Merge pull request #4 from clement0010/branch-A-CodeQuality

Add java docstring for all public methods

Commit:
Refactor code to support class type

Commit:
Add sort event and deadline feature

Commit:
Add filter tasks feature

clement0010 committed:
Merge pull request #5 from clement0010/branch-BCD-Extension

feat: Add category C extensions

Commit:
Add better gui to Duke

Commit:
Add product user guide

Commit:
Rename default filename for jar

Commit:
Fix unused import styling

clement0010 committed:
Update README.md
